### PR TITLE
Add "Sequencing side-effects in JAX" design note

### PR DESCRIPTION
**Pull request by sharadmv**
Adds a write up explaining how we sequence side-effectful computations in JAX.

### CODE REVIEW

File: docs/design_notes/sequencing_effects.md [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+## Motivation
+
+We'd like to support "ordered" effects. When we say ordered, we mean that the effects
+occur in the same order as we would if we were executing a single-threaded Python program.
+This is our main desideratum. In the presence of explicit parallelism like `pmap` or
+user threads, we don't need to maintain this behavior but at least if the user is not
+explicitly requesting parallelism, we'd like to support a single-threaded ordering.
+
+Before we dive into more, let's first step back and ask ourselves if it is okay
```
Comment (hawkinsp) [inline docs/design_notes/sequencing_effects.md:97]: Nit: "dive in more" or "dive into this more".

**Review by hawkinsp — APPROVED** [review]
This looks great!

File: docs/design_notes/sequencing_effects.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+Here we will actually start talking about implementation details. In practice, we'll need
+two separate types of tokens to sequence effects: one for each of the aforementioned sources
+of reordering. We'll need *runtime tokens* to sequence asynchronously dispatched 
+effectful computations and we'll need *compiler tokens* to sequence effects within computations.
+
+In practice, our computation will be rewritten to look like this:
+```python
+@jax.jit
+def f(runtime_token, x):
+  compiler_token = create_compiler_token()
```
[the remark this answers is not in the document]
Comment (sharadmv) [reply docs/design_notes/sequencing_effects.md:151]: I added an additional comment explaining why we can keep the runtime and compiler tokens "disconnected"

**Review by mattjj — APPROVED** [review]
This is _awesome_.

**Pull request merged**